vm_supervisor/ip.py

```diff
@@ -0,0 +1,530 @@
+                            },
+                            {"masquerade": None},
+                        ],
+                    }
+                }
+            }
+        ]
+
+        return self.execute_json_nft_commands(command)
+
+    def add_forward_rule_to_external(self, vm_id: int, interface: TapInterface) -> int:
+        """Creates a rule for the VM with the specified id to allow outbound traffic
+        Returns the exit code from executing the nftables commands"""
+        if vm_id not in self.vm_info:
+            logger.error(f"VM ruleset not yet initialized: {vm_id}")
+            return 1
+
+        command = [
+            {
+                "add": {
+                    "rule": {
+                        "family": "ip",
+                        "table": self.forwarding_info["table"],
+                        "chain": self.vm_info[vm_id]["filter_chain"],
+                        "expr": [
+                            {
+                                "match": {
+                                    "op": "==",
+                                    "left": {"meta": {"key": "iifname"}},
+                                    "right": interface.device_name,
+                                }
+                            },
+                            {
+                                "match": {
+                                    "op": "==",
+                                    "left": {"meta": {"key": "oifname"}},
+                                    "right": self.external_interface,
+                                }
+                            },
+                            {"accept": None},
+                        ],
+                    }
+                }
+            }
+        ]
+
+        return self.execute_json_nft_commands(command)
+
+    def setup_nftables_for_vm(self, vm_id: int, interface: TapInterface) -> None:
+        """Sets up chains for filter and nat purposes specific to this VM, and makes sure those chains are jumped to"""
+        if vm_id in self.vm_info:
+            logger.error(f"VM already setup in firewall: {vm_id}")
+            return
+
+        self.vm_info[vm_id] = {
+            "nat_chain": f"aleph-vm-nat-{vm_id}",
+            "filter_chain": f"aleph-vm-filter-{vm_id}",
+        }
+        self.add_postrouting_chain(self.vm_info[vm_id]["nat_chain"])
+        self.add_forward_chain(self.vm_info[vm_id]["filter_chain"])
+        self.add_masquerading_rule(vm_id, interface)
+        self.add_forward_rule_to_external(vm_id, interface)
+
+    def teardown_nftables_for_vm(self, vm_id: int) -> None:
+        """Remove all nftables rules related to the specified VM"""
+        if vm_id not in self.vm_info:
+            logger.error(f"No firewall configuration found to teardown for vm {vm_id}")
+            return
+
+        self.remove_chain(self.vm_info[vm_id]["nat_chain"])
+        self.remove_chain(self.vm_info[vm_id]["filter_chain"])
+
+
+class Network:
+    firewall: Firewall = Firewall()
```

Comment:
Initializing a variable in the class definition stores the instance on the class and not the instance.
This is usually not what we are looking for, especially when running tests.

```suggestion
    firewall: Firewall
    def __init__(self):
        self.firewall = Firewall()
```